### PR TITLE
 Group use declarations are incorrectly fixed by the PSR2 standard

### PR DESCRIPTION
**Pull request by jasonmccreary**

### CODE REVIEW

**Comment by jasonmccreary** [comment]
Per our discussion here https://github.com/squizlabs/PHP_CodeSniffer/issues/878, this is a first pass at a patch to *ungroup* grouped `use` statements when *fixing* for the PSR-2 coding standard.

While this passes a basic test, it would not work for *mixed* grouped `use` statements. For example:

```php
use some\ns {
    ClassA,
    const CON,
    function a_func
};
```

I'm not even sure such code is allowed in PHP 7. But I'm glad to adjust if so.

Also, the code could probably stand for some optimizations as I'm not as familiar with the API.

**Comment by gsherwood** [comment]
I think it looks pretty good, but I've got a couple of suggestions for you to consider.

1. It appears that mixed use group syntax is valid as per the RFC (https://wiki.php.net/rfc/group_use_declarations) so the sniff should handle fixing that. It doesn't look too hard given your code though.

2. It would be good to have tests for grouped `use function` and `use const` statements as well as the mixed group. Ripping them straight from the RFC would give this sample code:
```php
use function foo\math\{ sin, cos, cosh };
use const foo\math\{ PI, E, GAMMA, GOLDEN_RATIO };
use foo\math\{ Math, const PI, function sin, function cos, function cosh };
```
Which should be converted to:
```php
use function foo\math\sin;
use function foo\math\cos;
use function foo\math\cosh;
use const foo\math\PI
use const foo\math\E;
use const foo\math\GAMMA;
use const foo\math\GOLDEN_RATIO;
use foo\math\Math
use const foo\math\PI;
use function foo\math\sin;
use function foo\math\cos;
use function foo\math\cosh;
```

3. During fixing, it might be easier to just construct the new use statement lines, add that content before the `use` keyword, then wipe everything from the old use statement. This also ensures every token inside the grouped use statement is modified, so other sniffs wont get a go to change things until the next fixing loop - making conflicts impossible.

My general comment is: you've done a really good job on one of the most complex fixers. Thanks for the PR.

File: CodeSniffer/Standards/PSR2/Sniffs/Namespaces/UseDeclarationSniff.php [hunk 1]
```diff
@@ -68,14 +68,53 @@ public function process(PHP_CodeSniffer_File $phpcsFile, $stackPtr)
         }
 
         // Only one USE declaration allowed per statement.
-        $next = $phpcsFile->findNext(array(T_COMMA, T_SEMICOLON), ($stackPtr + 1));
-        if ($tokens[$next]['code'] === T_COMMA) {
+        $next = $phpcsFile->findNext(array(T_COMMA, T_SEMICOLON, T_OPEN_USE_GROUP), ($stackPtr + 1));
```
Comment (aik099) [inline CodeSniffer/Standards/PSR2/Sniffs/Namespaces/UseDeclarationSniff.php]: Is `T_OPEN_USE_GROUP` created also, when PHP 5 code is parsed OR PHP 5 is used to parse PHP 7 code?

Comment (gsherwood) [inline CodeSniffer/Standards/PSR2/Sniffs/Namespaces/UseDeclarationSniff.php]: Yes. It's a PHPCS-specific thing. PHP doesn't tokenise that code any differently, so PHPCS does it for all PHP versions.

**Comment by jasonmccreary** [comment]
Thanks. I will add tests for the mixed group `use` statements.

I thought about regenerating the `use` statements. What concerned me was something like the following:

```php
use some\ns\{
    // use whatever from here
    whatever,
};
```

Thoughts?

**Comment by jasonmccreary** [comment]
I have updated this to handle *mixed groups*.

I didn't rewrite it to rebuild the use statements as I felt that could get more complicated when attempting to account for things like comments. Nonetheless, open to suggestions as I'm sure this could be optimized.

**Comment by gsherwood** [comment]
> I opted not to rebuild the use statements as I felt that may be more complicated when trying to account for things like comments.

I didn't consider that. You've absolutely picked the right way.

**Comment by gsherwood** [comment]
This was a great PR. Thanks a lot.

**Comment by jasonmccreary** [comment]
Anytime! I use PHPCS on many projects. Glad I could give back.

**Pull request merged**